### PR TITLE
add `dyn ` to display of dynamic (trait) types

### LINKED ISSUE #49277: use `dyn` syntax in error messages
We have not yet stabilized `dyn Trait` (see https://github.com/rust-lang/rust/issues/49218), but once we do, we should start using it in error messages. So for example this test:

```rust
#![feature(dyn_trait)]
use std::fmt::{Debug, Display};
fn main() {
    let x: Box<dyn Debug> = Box::new(1);
    let _: Box<dyn Display> = x;
}
```

should print

```
error[E0308]: mismatched types
 --> src/main.rs:5:31
  |
5 |     let _: Box<dyn Display> = x;
  |                               ^ expected trait `std::fmt::Display`, found trait `std::fmt::Debug`
  |
  = note: expected type `std::boxed::Box<dyn std::fmt::Display>`
             found type `std::boxed::Box<dyn std::fmt::Debug>`
```

### PR DESCRIPTION
**Pull request by zackmdavis**
~~I'm not sure we want the `dyn` in the E0277 "trait bound [...] is not satisfied" messages ("bound" sounds like a different thing in contrast to the names of specific trait-object types like `Box<dyn Trait>`), but I'm finding the code I would need to change that hard to follow—the [display object seems to](https://github.com/rust-lang/rust/blob/f0805a4421449bd6fe3096d63820fbebe2bfcd1d/src/librustc/traits/error_reporting.rs#L600) be a [`Predicate::Trait`](https://github.com/rust-lang/rust/blob/f0805a4421449bd6fe3096d63820fbebe2bfcd1d/src/librustc/ty/mod.rs#L962) variant, whose [`Display` implementation](https://github.com/rust-lang/rust/blob/f0805a4421449bd6fe3096d63820fbebe2bfcd1d/src/librustc/util/ppaux.rs#L1309) calls `.print` on its `PolyTraitPredicate` member, [which is a type alias](https://github.com/rust-lang/rust/blob/f0805a4421449bd6fe3096d63820fbebe2bfcd1d/src/librustc/ty/mod.rs#L1112) for `ty::Binder<TraitPredicate<'tcx>>`, whose [`Display` implementation](https://github.com/rust-lang/rust/blob/f0805a4421449bd6fe3096d63820fbebe2bfcd1d/src/librustc/util/ppaux.rs#L975-L985) ... _&c._— so maybe it's time to pull-request this and see what reviewers think.~~

 Resolves #49277 (?).

r? @nikomatsakis

### CODE REVIEW

**Comment by sanxiyn** [comment]
Legit test failures.

**Comment by nikomatsakis** [comment]
@zackmdavis (btw, `--bless` may be very useful in keeping this PR up to date)

File: src/librustc/util/ppaux.rs [hunk 1]
```diff
@@ -1063,6 +1063,7 @@ define_print! {
                 TyParam(ref param_ty) => write!(f, "{}", param_ty),
                 TyAdt(def, substs) => cx.parameterized(f, substs, def.did, &[]),
                 TyDynamic(data, r) => {
+                    write!(f, "dyn ")?;
```
Comment (nikomatsakis) [inline src/librustc/util/ppaux.rs]: Hmm, I think this may do the wrong thing in terms of precedence in some cases. For example, `&dyn Foo + 'a` needs some parens, right? I'm not 100% sure how this is handled already, since the same applies to `&Foo + 'a`... it kind of looks like we just get it wrong now?

Maybe we can concoct a test case to demonstrate this...

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
This seems... basically right? The main thing I see is the precedence issue, but perhaps I am overlooking some subtlety? (cc @eddyb @petrochenkov)

**Comment by nikomatsakis** [comment]
Oh, I sort of missed this:

> I'm not sure we want the dyn in the E0277 "trait bound [...] is not satisfied" messages 

Or at least forgot about it. Let me look at that specifically.

**Comment by eddyb** [comment]
> but I'm finding the code I would need to change that hard to follow

So, all of that eventually ends up probably in the `TraitRef` implementation, but we don't want `dyn` there, since trait bounds are different from dynamic trait objects. You can't write `T: dyn Trait`.

**Comment by nikomatsakis** [comment]
Right, right. Let me take a look. I forget how it works now; I know I have had at least one branch that tried to refactor *precisely this point* but I think it never landed (was part of a lazy norm push on my part)...

File: src/test/ui/const-unsized.stderr [hunk 2]
```diff
@@ -16,13 +16,13 @@ LL | const CONST_FOO: str = *"foo";
    = help: the trait `std::marker::Sized` is not implemented for `str`
    = note: constant expressions must have a statically known size
 
-error[E0277]: the trait bound `std::fmt::Debug + std::marker::Sync + 'static: std::marker::Sized` is not satisfied
+error[E0277]: the trait bound `dyn std::fmt::Debug + std::marker::Sync + 'static: std::marker::Sized` is not satisfied
```
Comment (nikomatsakis) [inline src/test/ui/const-unsized.stderr]: I have to admit — tracing through ppaux.rs — I couldn't quite tel how this arises. Can somebody clarify for me? I imagine that this message is produced from `traits/error_reporting.rs`:

https://github.com/rust-lang/rust/blob/90f34b5f83d8fb0dd39de7a6d7cfac4d10cb2218/src/librustc/traits/error_reporting.rs#L598-L601

But it seems like that is printing a `PolyTraitRef`, and I'm not clear on the path that leads from that code to the `TyDynamic` code. What am I missing?

Comment (eddyb) [reply src/test/ui/const-unsized.stderr]: The `TyDynamic` is the `Self` type, printed by this:
https://github.com/rust-lang/rust/blob/4a9c58c6bd0e609538053a35685e206f66019820/src/librustc/util/ppaux.rs#L1258

Comment (nikomatsakis) [reply src/test/ui/const-unsized.stderr]: Oh; in that case, this seems...correct.

Comment (nikomatsakis) [reply src/test/ui/const-unsized.stderr]: it might be nicer to print "the trait `Sized` is not implemented for `dyn ...`", like we used to, but that seems separate

**Comment by zackmdavis** [comment]
> so specifically which cases were worrying you?

The message that says "the trait bound `dyn MyTrait`" when (as Eddy [pointed out](https://github.com/rust-lang/rust/pull/51104#issuecomment-393599638)) you can't use `T: dyn MyTrait` as a bound. I like [your idea of rephrasing the message](https://github.com/rust-lang/rust/pull/51104#discussion_r193553874) (but I won't have time to address this for at least a few days).

**Comment by eddyb** [comment]
@zackmdavis IMO it's fine, since "the trait bound `dyn Foo: Bar` is not satisfied" means `dyn Foo` doesn't implement `Bar`, i.e. `dyn Foo` isn't "the trait bound", it's what the bound is "applied to".

**Comment by zackmdavis** [comment]
Oh, right! :flushed: OK, so then the only things left to do are:

 - [ ] write a test and possibly fix the [suspected](https://github.com/rust-lang/rust/pull/51104#discussion_r192163071) parentheses precedence issue
 -  [ ] fix the remaining test expectations

—which I should get to early next week.

**Comment by pietroalbini** [comment]
[note for future triage: the author has not so much time for rust at the moment, let's keep this open though https://github.com/rust-lang/rust/pull/51149#issuecomment-396127722]

**Comment by zackmdavis** [comment]
@nikomatsakis updated to add parentheses when there's a `+ 'lifetime`, which [the parser seems to require](https://github.com/rust-lang/rust/blob/e11c95dda154762276b92184df10e942991594ec/src/libsyntax/parse/parser.rs#L1582-L1588). (No dedicated test because this affects a lot of the existing tests.)

**Comment by zackmdavis** [comment]
@nikomatsakis rebased :checkered_flag:
